File: 73-set-matrix-zeroes/set-matrix-zeroes.py
class Solution(object):
    def setZeroes(self, arr):
        n = len(arr)
        m = len(arr[0])
        col1 = 1

        # Step 1: mark rows & columns
        for i in range(n):
            for j in range(m):
                if arr[i][j] == 0:
                    arr[i][0] = 0
                    if j != 0:
                        arr[0][j] = 0
                    else:
                        col1 = 0

        # Step 2: update inner matrix
        for i in range(1, n):
            for j in range(1, m):
                if arr[i][0] == 0 or arr[0][j] == 0:
                    arr[i][j] = 0

        # Step 3: handle first row
        if arr[0][0] == 0:
            for j in range(m):
                arr[0][j] = 0

        # Step 4: handle first column
        if col1 == 0:
            for i in range(n):
                arr[i][0] = 0

File: 73-set-matrix-zeroes/set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& nums) {
        int n = nums.size();
        int m =nums[0].size();
        int col1 =1;
        for(int i =0;i<n;i++){
            for(int j=0;j<m;j++){
                if(nums[i][j]==0){
                    nums[i][0]=0;
                    if(j!=0){
                        nums[0][j]=0;
                    }
                    else col1 =0;
                }
            }
        }
        for(int i =1;i<n;i++){
            for(int j=1 ;j<m;j++){
                if(nums[i][j]!=0){
                    if(nums[0][j]==0 || nums[i][0]==0){
                        nums[i][j]=0;
                    }
                }
            }
        }
        if(nums[0][0]==0){
            for(int j =0;j<m;j++){
                nums[0][j]=0;
            }
        }
        if(col1 ==0){
            for(int i =0;i<n;i++){
                nums[i][0]=0;
            }
        }
    }
};